packages/cells/src/widget.ts

```diff
@@ -1419,6 +1433,67 @@ export class MarkdownCell extends AttachmentsCell<IMarkdownCellModel> {
     return this._ready.promise;
   }
 
+  /**
+   * Text that represents the heading if cell is a heading.
+   * Returns empty string if not a heading.
+   */
+  get headingInfo(): { text: string; level: number } {
+    let text = this.model.value.text;
+    const lines = text.split('\n');
+    const line = lines[0];
+    const line2 = lines.length > 1 ? lines[1] : undefined;
+
+    let match = line.match(/^([#]{1,6}) (.*)/);
+    let match2 = line2 && line2.match(/^([=]{2,}|[-]{2,})/);
+    let match3 = line.match(/<h([1-6])(.*)>(.*)<\/h\1>/i);
```

Comment:
This regexp will fail if there are multiple heading of this level in the same cell I think.

Reply:
I think that this regexp should be lazy, this is: `<h([1-6])(.*?)>(.*?)<\/h\1>` as otherwise the following:

`<h1>a</h1><h1>b</h1>` would match group 2 = `>a</h1><h1` and group 3 = `b`. To additionally prevent ReDoS we would need something more fancy like `<h([1-6])(?=([^>]*>))\2(?=(.*?<\/h\1>))\3` which would require additional processing of group 2 to remove `>` and group 3 to remove `</h[1-6]>` suffixes. I can imagine that it will not resolve all possible ReDoS scenarios as parsing HTML with regular expressions is expensive (and not generally feasible nor useful).

Given that [marked.js](https://github.com/markedjs/marked) is already a dependency (which is tested against ReDoS) that actually does display the markdwon, could we use it instead to parse the markdown and extract the headers? See "Access to lexer and parser" in https://marked.js.org/using_pro#parser

Reply:
ReDoS aside, using marked would guarantee consistency with the headers that do get display in the cells.

Reply:
Oh I didn't realize that - I can add in the marked